core-bundle/src/Controller/Page/RootPageController.php

```diff
@@ -39,8 +39,8 @@ private function getNextPage(int $rootPageId): PageModel
             return $nextPage;
         }
 
-        if ($this->has('logger')) {
-            $this->get('logger')->error(
+        if ($this->container->has('logger')) {
+            $this->container->get('logger')->error(
```

Comment:
I really think this should have been a constructor injection.

Reply:
Everything should be constructor injection if it does not need to be conditionally loaded. E.g. every controller that uses "service A" all the time, should use constructor injection and never subscribe to it. You should only subscribe if you provide e.g. abstract functions other children *can* use. Just like `AbstractController` itself does. There are other ideas like `AbstractJobType` etc. But if the code needs the service, it should always use constructor injection.
Probably not in this PR though ;)